docs/src/main/tut/kleisli.md

```diff
@@ -0,0 +1,261 @@
+scaladoc: "#cats.data.Kleisli"
+---
+# Kleisli
+Kleisli is a data type that will come in handy often, especially if you are working with monadic functions.
+
+## Functions
+One of the most useful properties of functions is that they **compose**. That is, given a function
+`A => B` and a function `B => C`, we can combine them to create a new function `A => C`. It is through
+this compositional property that we are able to write many small functions and compose them together
+to create a larger one that suits our needs.
+
+Sometimes, our functions will need to return monadic values. For instance, consider the following set of functions.
+
+```tut
+val parse: String => Option[Int] = s =>
+  try {
+    Some(s.toInt)
+  } catch {
+    case _: NumberFormatException => None
+  }
+
+val reciprocal: Int => Option[Double] =
+  i => if (i != 0) Some(1.0 / i) else None
+```
+
+As it stands we cannot use `Function1.compose` (or `Function1.andThen`) to compose these two functions.
+The output type of `parse` is `Option[Int]` whereas the input type of `reciprocal` is `Int`.
+
+This is where `Kleisli` comes into play.
+
+## Kleisli
+At it's core, `Kleisli[F[_], A, B]` is just a wrapper around the function `A => F[B]`. Depending on the
+properties of the `F[_]`, we can do different things with `Kleisli`s. For instance, if `F[_]` has a
+`FlatMap[F]` instance (we can call `flatMap` on `F[A]` values), we can
+compose two `Kleisli`s much like we can two functions.
+
+```tut
+import cats.FlatMap
+
+final case class Kleisli[F[_], A, B](run: A => F[B]) {
+  def compose[Z](k: Kleisli[F, Z, A])(implicit F: FlatMap[F]): Kleisli[F, Z, B] =
+    Kleisli[F, Z, B](z => F.flatMap(k.run(z))(run))
+}
+```
+
+Returning to our earlier example:
+
+```tut
+// Bring in cats.FlatMap[Option] instance
+import cats.std.option._
+
+val parse = Kleisli((s: String) => try { Some(s.toInt) } catch { case _: NumberFormatException => None })
+
+val reciprocal = Kleisli((i: Int) => if (i == 0) None else Some(1.0 / i))
+
+val parseAndReciprocal = reciprocal.compose(parse)
+```
+
+`Kleisli#andThen` can be defined similarly.
+
+It is important to note that the `F[_]` having a `FlatMap` (or a `Monad`) instance is not a hard requirement -
+we can do useful things with weaker requirements. Such an example would be `Kleisli#map`, which only requires
+that `F[_]` have a `Functor` instance (e.g. is equipped with `map: F[A] => (A => B) => F[B]`).
+
+```tut
+import cats.Functor
+
+final case class Kleisli[F[_], A, B](run: A => F[B]) {
+  def map[C](f: B => C)(implicit F: Functor[F]): Kleisli[F, A, C] =
+    Kleisli[F, A, C](a => F.map(run(a))(f))
+}
+```
+
+Below are some more methods on `Kleisli` that can be used so long as the constraint on `F[_]`
+is satisfied.
+
+Method    | Constraint on `F[_]`
+--------- | -------------------
+andThen   | FlatMap
+compose   | FlatMap
+flatMap   | FlatMap
+lower     | Monad
+map       | Functor
+traverse  | Applicative
+
```

Comment:
Might be nice to put the signatures in a column, if there's room ... `lower` and `traverse` are likely to be unfamiliar.